kuka_eki_hw_interface/src/kuka_eki_hw_interface.cpp

```
@@ -0,0 +1,204 @@
+  }
+
+  // Create ros_control interfaces (joint state and position joint for all dof's)
+  for (std::size_t i = 0; i < n_dof_; ++i)
+  {
+    // Joint state interface
+    joint_state_interface_.registerHandle(
+        hardware_interface::JointStateHandle(joint_names_[i], &joint_position_[i], &joint_velocity_[i],
+                                             &joint_effort_[i]));
+
+    // Joint position control interface
+    position_joint_interface_.registerHandle(
+        hardware_interface::JointHandle(joint_state_interface_.getHandle(joint_names_[i]),
+                                        &joint_position_command_[i]));
+  }
+
+  // Register interfaces
+  registerInterface(&joint_state_interface_);
+  registerInterface(&position_joint_interface_);
+
+  ROS_INFO_STREAM_NAMED("kukw_eki_hw_interface", "Loaded Kuka EKI hardware interface");
+}
+
+
+KukaEkiHardwareInterface::~KukaEkiHardwareInterface()
+{
+
+}
+
+
+bool KukaEkiHardwareInterface::socket_read_state(std::vector<double> &joint_position)
+{
+  // TODO(BrettHemes): Look into making receive non-blocking?
+  boost::array<char, 2048> in_buffer;
+
+  // Read socket buffer
+  size_t len = eki_state_socket_.receive_from(boost::asio::buffer(in_buffer), eki_state_endpoint_);
+
+  // Update joint positions from XML packet (if received)
+  if (len > 0)
+  {
+    // Parse XML
+    TiXmlDocument xml_in;
+    in_buffer[len] = '\0';  // null-terminate data buffer for parsing (expects c-string)
+    xml_in.Parse(in_buffer.data());
+    TiXmlElement* robot_state = xml_in.FirstChildElement("RobotState");
+
+    // Extract axis positions
+    double joint_pos_deg;
+    char axis_name[] = "A1";
+    for (int i = 0; i < n_dof_; ++i)
+    {
+      robot_state->Attribute(axis_name, &joint_pos_deg);
+      joint_position[i] = DEG2RAD * joint_pos_deg;
+      axis_name[1]++;
+    }
+
+    return true;
+  }
+
+  return false;
```

I'm a big fan of 'return/fail early'.

Perhaps doing something like the following would increase readability:

```c++
...

size_t len = eki_state_socket_.receive_from(..);

if (len == 0)
  return false;

// continue with received data
...

return true;
```